xnmt/decorators.py

```diff
@@ -8,7 +6,8 @@ def recursive(f):
   '''
   # Wrapper to the method to call recursively, instead of invoking only f()
   def rec_f(obj, *args, **kwargs):
-    assert(issubclass(obj.__class__, model.HierarchicalModel))
+    from model import HierarchicalModel
```

Comment:
Why are we importing classes in the middle of the code? Does this mean that this import will be called over and over again?

Reply:
I did this to fix problems with a circular import: previously, model.py imported decorators.py, and decorators.py imported model.py.

Python will only execute the first import, so it should be okay, but would probably be even better to improve the structure so we don't need to worry about this?

Reply:
right. I think the better structure will be putting the HierarchicalModel and all of the decorators together in the same file (hier_model.py). It is up to you though if you want to change it that way, or can be done in a separate pull request.

Reply:
Sure, let's do it one a later PR then! Part of this PR is uncovering a few structural issues, for example this one or also some things in xnmt_train.py which I noticed while writing tests but won't be able to fix myself right now.